test/distributed/_composable/fsdp/test_fully_shard_state_dict.py

```diff
@@ -0,0 +1,157 @@
+        model = nn.Sequential(*[MLP(mlp_dim) for _ in range(3)])
+        model = parallelize_module(
+            model,
+            device_mesh=tp_mesh,
+            parallelize_plan={
+                "0.in_proj": ColwiseParallel(),
+                "0.out_proj": RowwiseParallel(),
+                "1.in_proj": ColwiseParallel(),
+                "1.out_proj": RowwiseParallel(),
+                "2.in_proj": ColwiseParallel(),
+                "2.out_proj": RowwiseParallel(),
+            },
+        )
+        for mlp in model:
+            fully_shard(mlp, mesh=dp_mesh)
+        fully_shard(model, mesh=dp_mesh)
+        self._test_state_dict_save_load(model)
+
+    def _test_state_dict_save_load(self, model: nn.Module):
+        for param_name, param in model.named_parameters():
+            self.assertIsInstance(
+                param,
+                DTensor,
+                f"Expects parameters to be sharded as DTensors but got {param_name} "
+                f"as {type(param)}: {param}",
+            )
+        old_fill_value = 1
+        new_fill_value = 42 + self.rank
+        with torch.no_grad():
+            for param in model.parameters():
+                param.fill_(old_fill_value)
+        # Use that the parameters are currently sharded, meaning that their
+        # data pointers correspond to the sharded parameter data
+        param_name_to_data_ptr = {
+            n: p._local_tensor.data_ptr() for n, p in model.named_parameters()
+        }
+        ref_sharded_sizes = [p.size() for p in model.parameters()]
+        state_dict = model.state_dict()
+        for param, ref_sharded_size in zip(model.parameters(), ref_sharded_sizes):
+            self.assertEqual(param.size(), ref_sharded_size)
+            self.assertTrue(isinstance(param, nn.Parameter))
+
+        # Verify that keys match, values are DTensors, and values share the
+        # same storage as the existing sharded parameter data
+        self.assertEqual(set(state_dict.keys()), set(param_name_to_data_ptr.keys()))
+        for param_name, tensor in state_dict.items():
+            self.assertTrue(isinstance(tensor, DTensor))
+            if param_name_to_data_ptr[param_name] == 0:
+                # Check that this is padding (added by DTensor)
+                self.assertGreater(self.rank, 0)
+                self.assertEqual(torch.count_nonzero(tensor._local_tensor).item(), 0)
+            else:
+                self.assertEqual(
+                    tensor._local_tensor.data_ptr(), param_name_to_data_ptr[param_name]
+                )
+
+        # Verify that we can load a new state dict that contains DTensors with
+        # storages different from the current model parameters
+        new_state_dict: Dict[str, DTensor] = {}
+        for param_name, dtensor in state_dict.items():
+            # Construct new DTensors to exercise load state dict writeback
+            new_state_dict[param_name] = dtensor.detach().clone().fill_(new_fill_value)
```

Comment:
I think this is essentially testing the same sharded parameter can be saved/loaded in the same way? i.e. load state_dict essentially just load back a manually written value. 

This is working because **both** 2D state_dict and the new model state_dict generates the same strided sharding, so directly loading with the same topology works.

IMO for 2d checkpoint load re-sharding, it does not work, suppose we have a module that have only one parameter:

```
param = torch.arange(4) + 5
```
after TP (assuming TP Shard(0)), it become:
```
rank 0, 2: [5, 6]
rank 1, 3: [7, 8]
```
Then after FSDP it would become:
```
rank 0: [5]
rank 1: [7]
rank 2: [6]
rank 3: [8]
```
DCP would write offsets as if it would be:
```
rank 0: [5]
rank 1: [6]
rank 2: [7]
rank 3: [8]
```
Then resharding would incorrectly load the data to the wrong offset

cc @wz337